compiler/optimizer/OSRDefAnalysis.hpp

```diff
@@ -99,19 +99,30 @@ class TR_OSRLiveRangeAnalysis : public TR::Optimization
 
    private:
 
-   bool canAffordAnalysis();
+   bool shouldPerformAnalysis();
+   int32_t fullAnalysis(bool sharedParm, bool containsPendingPush);
+   int32_t partialAnalysis();
+
    void buildOSRLiveRangeInfo(TR::Node *node, TR_BitVector *liveVars, TR_OSRPoint *osrPoint,
-      int32_t *liveLocalIndexToSymRefNumberMap, int32_t maxSymRefNumber, int32_t numBits,
-      TR_OSRMethodData *osrMethodData);
+      int32_t *liveLocalIndexToSymRefNumberMap, int32_t numBits, TR_OSRMethodData *osrMethodData);
    void buildOSRSlotSharingInfo(TR::Node *node, TR_BitVector *liveVars, TR_OSRPoint *osrPoint,
-      int32_t *liveLocalIndexToSymRefNumberMap, TR_BitVector *slotSharingVars);
+      int32_t *liveLocalIndexToSymRefNumberMap, TR_BitVector &slotSharingVars);
+
+   void pendingPushLiveRangeInfo(TR::Node *node, TR_BitVector *liveSymRefs,
+      TR_BitVector &allPendingPushSymRefs, TR_OSRPoint *osrPoint, TR_OSRMethodData *osrMethodData);
+   void pendingPushSlotSharingInfo(TR::Node *node, TR_BitVector *liveSymRefs,
+      TR_BitVector &slotSharingSymRefs, TR_OSRPoint *osrPoint);
+
    void maintainLiveness(TR::Node *node, TR::Node *parent, int32_t childNum, vcount_t  visitCount,
        TR_Liveness *liveLocals, TR_BitVector *liveVars, TR::Block *block);
    TR::TreeTop *collectPendingPush(TR_ByteCodeInfo bci, TR::TreeTop *pps, TR_BitVector *liveVars);
 
    TR_BitVector *_deadVars;
    TR_BitVector *_liveVars;
-   TR_BitVector *_pendingPushVars;
+
+   TR_BitVector _sharedSymRefs;
+   TR_BitVector _pendingPushSymRefs;
+   TR_BitVector _workBitVector;
```

Comment:
Can you provide a justification for why these need to be heap allocated and not stack allocated like _deadVars and _liveVars? This would seem to be almost an effective 'leak' of memory since these are only live in the optimization...

Reply:
There was no suitable justification. They have been moved to the stack. Moreover, `workBitVector` has been used to replace `deadVars`.